src/backend/lib/utils.js

```diff
@@ -0,0 +1,268 @@
+ * @returns {boolean}
+ */
+function isUpgradeCandidate({ deviceInfo, installer, currentApkInfo }) {
+  // TODO: 64 bit devices can run 32 bit APKs
+  const isSupportedAbi = !!installer.arch.find(arch =>
+    deviceInfo.supportedAbis.includes(arch)
+  );
+  if (!isSupportedAbi) return false;
+  if (installer.minSdkVersion > deviceInfo.sdkVersion) return false;
+  if (installer.applicationId !== currentApkInfo.applicationId) return false;
+  if (installer.platform !== "android") return false;
+  if (installer.versionCode < currentApkInfo.versionCode) return false;
+  if (!semver.valid(installer.versionName)) return false;
+  // TODO: Check for internal and release candidate builds e.g. a release
+  // candidate is only an upgrade candidate if the current apk is a release
+  // candidate
+  return true;
+}
+
+/**
+ * From a list of installers, return the most recent compatible upgrade (if
+ * there is one)
+ *
+ * @template {InstallerInt | InstallerExt} T
+ * @param {object} options
+ * @param {import('./types').DeviceInfo} options.deviceInfo
+ * @param {Array<T>} options.installers
+ * @param {InstallerInt} options.currentApkInfo
+ * @returns {T}
+ */
+function getUpgradeCandidate({ deviceInfo, installers, currentApkInfo }) {
+  const upgradeCandidates = installers.filter(installer =>
+    isUpgradeCandidate({
+      installer,
+      currentApkInfo,
+      deviceInfo,
+    })
+  );
+  return upgradeCandidates.sort(installerCompare).reverse()[0];
+}
+
+/**
+ * For a given filepath to an APK, get the UpgradeOption details
+ *
+ * @param {string} filepath
+ * @returns {Promise<InstallerInt>}
+ */
+async function getInstallerInfo(filepath) {
+  if (path.extname(filepath) !== ".apk")
+    throw new Error(`${filepath} is not an .apk file`);
+  const [
+    hash,
+    stat,
+    { versionName, versionCode, applicationId, arch, minSdkVersion },
+  ] = await Promise.all([
+    hasha.fromFile(filepath, { algorithm: "sha256" }),
+    fs.promises.stat(filepath),
+    getApkMetadata(filepath),
+  ]);
+  return {
+    hash,
+    hashType: "sha256",
+    versionName,
+    versionCode,
+    applicationId,
+    platform: "android",
+    arch,
+    size: stat.size,
+    minSdkVersion,
+    filepath,
+  };
+}
+
+/**
+ * Get version name, version code & applicationId from an apk file
+ *
+ * @param {string} filepath path to apk
+ * @returns {Promise<Pick<InstallerInt, 'versionName' | 'versionCode' | 'applicationId' | 'arch' | 'minSdkVersion'>>}
+ */
+async function getApkMetadata(filepath) {
+  const parser = new ApkParser(filepath);
+  const [result, arch] = await Promise.all([
+    parser.parse(),
+    getApkAbis(filepath),
+  ]);
+  return {
+    versionName: result.versionName,
+    versionCode: result.versionCode,
+    applicationId: result.package,
+    arch,
+    minSdkVersion: (result.usesSdk || {}).minSdkVersion || 1,
+  };
+}
+
+// If an APK does not have native code (e.g. no `lib` folder) then we assume it runs on these ABIs
```

Comment:
Yes in my testing, if I create an empty Android app with no C++ code (eg only Java or Kotlin) then there is no "lib" folder. I assume that in that case the apk would run on any platform that Android runs on, but the main 4 (arm and x86, 32 and 64 bit) I think cover all likely platforms. In addition I don't see a future where we would ever release mapeo without native compiled code (it will either be node or rust I think)